lib/SILOptimizer/LoopTransforms/LICM.cpp

```diff
@@ -737,14 +819,185 @@ bool LoopTreeOptimization::optimizeLoop(
+
+  // Propagate the store-is-not-alive flag through the control flow in the loop,
+  // starting at the header.
+  SmallPtrSet<SILBasicBlock *, 16> storesNotAlive;
+  storesNotAlive.insert(header);
+  bool changed = false;
+  do {
+    changed = false;
+    for (SILBasicBlock *block : loop->blocks()) {
+      bool storeAlive = (storesNotAlive.count(block) == 0);
+      if (storeAlive && stores.count(block) == 0 &&
+          std::any_of(block->pred_begin(), block->pred_end(),
+            [&](SILBasicBlock *b) { return storesNotAlive.count(b) != 0; })) {
+        storesNotAlive.insert(block);
+        changed = true;
+      }
+    }
+  } while (changed);
+
+  auto isUnreachableBlock = [](SILBasicBlock *succ) {
+    return isa<UnreachableInst>(succ->getTerminator());
+  };
+
+  // Check if the store-is-not-alive flag reaches any of the exits.
+  for (SILBasicBlock *eb : exitingBlocks) {
+    // Ignore loop exits to blocks which end in an unreachable.
+    if (!std::any_of(eb->succ_begin(), eb->succ_end(), isUnreachableBlock) &&
+        storesNotAlive.count(eb) != 0) {
+      return false;
+    }
+  }
+  return true;
+}
+
+void LoopTreeOptimization::hoistLoadsAndStores(SILValue addr, SILLoop *loop, InstVector &toDelete) {
+
+  SmallVector<SILBasicBlock *, 4> exitingBlocks;
+  loop->getExitingBlocks(exitingBlocks);
+
+  // This is not a requirement for functional correctness, but we don't want to
+  // _speculatively_ load and store the value (outside of the loop).
+  if (!storesCommonlyDominateLoopExits(addr, loop, exitingBlocks))
+    return;
+
+  // Inserting the stores requires the exit edges to be not critical.
+  for (SILBasicBlock *exitingBlock : exitingBlocks) {
+    auto successors = exitingBlock->getSuccessors();
+    for (unsigned idx = 0, e = successors.size(); idx != e; ++idx) {
+      SILBasicBlock *succ = successors[idx];
+      if (!loop->contains(succ))
+        splitCriticalEdge(exitingBlock->getTerminator(), idx, DomTree, LoopInfo);
+    }
+  }
+
+  SILBasicBlock *preheader = loop->getLoopPreheader();
+  assert(preheader && "Expected a preheader");
+
+  // Initially load the value in the loop pre header.
+  SILBuilder B(preheader->getTerminator());
+  auto *initialLoad = B.createLoad(preheader->getTerminator()->getLoc(), addr,
+                                   LoadOwnershipQualifier::Unqualified);
+
+  SILSSAUpdater ssaUpdater;
+  ssaUpdater.Initialize(initialLoad->getType());
+  ssaUpdater.AddAvailableValue(preheader, initialLoad);
+
+  Optional<SILLocation> loc;
+  for (SILInstruction *I : LoadsAndStores) {
+    if (auto *SI = isStoreToAddr(I, addr)) {
+      // If there are multiple stores in a block, only the last one counts.
+      loc = SI->getLoc();
+      ssaUpdater.AddAvailableValue(SI->getParent(), SI->getSrc());
+    }
+  }
+
+  // Remove all stores and replace the loads with the current value.
+  SILBasicBlock *currentBlock = nullptr;
+  SILValue currentVal;
+  for (SILInstruction *I : LoadsAndStores) {
```

Comment:
This has an assumption that LoadsAndStores are in program order but there's no contract. It isn't specified at the declaration or when the LoadsAndStores list is created.